File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Consent/Scope.cs
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;

namespace Iaiso.Consent;

/// <summary>A verified consent scope, ready to attach to an execution.</summary>
public sealed class Scope
{
    public string Token { get; }
    public string Subject { get; }
    public IReadOnlyList<string> Scopes { get; }
    public string? ExecutionId { get; }
    public string Jti { get; }
    public long IssuedAt { get; }
    public long ExpiresAt { get; }
    public JsonObject Metadata { get; }

    public Scope(string token, string subject, IReadOnlyList<string> scopes,
                 string? executionId, string jti, long issuedAt, long expiresAt,
                 JsonObject? metadata)
    {
        Token = token;
        Subject = subject;
        Scopes = scopes;
        ExecutionId = executionId;
        Jti = jti;
        IssuedAt = issuedAt;
        ExpiresAt = expiresAt;
        Metadata = metadata ?? new JsonObject();
    }

    public bool Grants(string requested) => Iaiso.Consent.Scopes.Granted(Scopes, requested);

    public void Require(string requested)
    {
        if (!Grants(requested))
        {
            throw new InsufficientScopeException(Scopes, requested);
        }
    }
}

/// <summary>
/// In-memory revocation list. Production deployments should back this
/// with Redis, a database, or similar.
/// </summary>
public sealed class RevocationList
{
    private readonly HashSet<string> _revoked = new();
    private readonly ReaderWriterLockSlim _lock = new();

    public void Revoke(string jti)
    {
        _lock.EnterWriteLock();
        try { _revoked.Add(jti); }
        finally { _lock.ExitWriteLock(); }
    }

    public bool IsRevoked(string jti)
    {
        _lock.EnterReadLock();
        try { return _revoked.Contains(jti); }
        finally { _lock.ExitReadLock(); }
    }

    public int Count
    {
        get
        {
            _lock.EnterReadLock();
            try { return _revoked.Count; }
            finally { _lock.ExitReadLock(); }
        }
    }

    public IReadOnlySet<string> Snapshot()
    {
        _lock.EnterReadLock();
        try { return new HashSet<string>(_revoked); }
        finally { _lock.ExitReadLock(); }
    }
}

File: IAIso-v5.0/core/iaiso-java/iaiso-consent/src/main/java/io/iaiso/consent/Scope.java
package io.iaiso.consent;

import com.google.gson.JsonObject;

import java.util.Collections;
import java.util.List;

/** A verified consent scope, ready to attach to an execution. */
public final class Scope {
    private final String token;
    private final String subject;
    private final List<String> scopes;
    private final String executionId;   // nullable
    private final String jti;
    private final long issuedAt;
    private final long expiresAt;
    private final JsonObject metadata;

    public Scope(String token, String subject, List<String> scopes,
                 String executionId, String jti, long issuedAt, long expiresAt,
                 JsonObject metadata) {
        this.token = token;
        this.subject = subject;
        this.scopes = Collections.unmodifiableList(scopes);
        this.executionId = executionId;
        this.jti = jti;
        this.issuedAt = issuedAt;
        this.expiresAt = expiresAt;
        this.metadata = metadata != null ? metadata : new JsonObject();
    }

    public String getToken() { return token; }
    public String getSubject() { return subject; }
    public List<String> getScopes() { return scopes; }
    public String getExecutionId() { return executionId; }
    public String getJti() { return jti; }
    public long getIssuedAt() { return issuedAt; }
    public long getExpiresAt() { return expiresAt; }
    public JsonObject getMetadata() { return metadata; }

    /** Does the scope grant the requested string? */
    public boolean grants(String requested) {
        return Scopes.granted(scopes, requested);
    }

    /** Throws {@link InsufficientScopeException} if not granted. */
    public void require(String requested) {
        if (!grants(requested)) {
            throw new InsufficientScopeException(scopes, requested);
        }
    }
}
